awspub/tests/test_sns.py

```diff
@@ -0,0 +1,88 @@
+import pathlib
+from unittest.mock import patch
+
+import pytest
+
+from awspub import context, exceptions, sns
+
+curdir = pathlib.Path(__file__).parent.resolve()
+
+
+@pytest.mark.parametrize(
+    "imagename,called_sns_publish, publish_call_count",
+    [
+        ("test-image-10", True, 1),
+        ("test-image-11", True, 4),
+    ],
+)
+def test_sns_publish(imagename, called_sns_publish, publish_call_count):
+    """
+    Test the send_notification logic
+    """
+    with patch("boto3.client") as bclient_mock:
+        instance = bclient_mock.return_value
+        ctx = context.Context(curdir / "fixtures/config1.yaml", None)
+        image_conf = ctx.conf["images"][imagename]
+
+        for region in image_conf["regions"]:
+            sns.SNSNotification(ctx, imagename, region).publish()
+
+        assert instance.publish.called == called_sns_publish
+        assert instance.publish.call_count == publish_call_count
+
+
+@pytest.mark.parametrize(
+    "imagename",
+    [
+        ("test-image-10"),
+        ("test-image-11"),
+    ],
+)
+def test_sns_publish_fail(imagename):
+    """
+    Test the send_notification logic
+    """
+    with patch("boto3.client") as bclient_mock:
+        instance = bclient_mock.return_value
+        ctx = context.Context(curdir / "fixtures/config1.yaml", None)
+        image_conf = ctx.conf["images"][imagename]
+
+        # topic2 is invalid topic
+        def side_effect(*args, **kwargs):
+            topic_arn = kwargs.get("TopicArn")
+            if "topic1" in topic_arn:
+                raise Exception(
+                    "An error occurred (NotFound) when calling the Publish operation: " "Topic does not exist."
+                )
+
+        instance.publish.side_effect = side_effect
+
+        for region in image_conf["regions"]:
+            with pytest.raises(exceptions.AWSNotificationException):
+                sns.SNSNotification(ctx, imagename, region).publish()
+
+
+@pytest.mark.parametrize(
+    "imagename, partition",
+    [
+        ("test-image-10", "aws-cn"),
+        ("test-image-11", "aws"),
+    ],
+)
+def test_sns__get_topic_arn(imagename, partition):
+    """
+    Test the send_notification logic
+    """
+    with patch("boto3.client") as bclient_mock:
+        instance = bclient_mock.return_value
+        ctx = context.Context(curdir / "fixtures/config1.yaml", None)
+        image_conf = ctx.conf["images"][imagename]
+
+        instance.get_caller_identity.return_value = {"Account": "1234", "Arn": f"arn:{partition}:iam::1234:user/test"}
+
+        for region in image_conf["regions"]:
+            for topic in image_conf["sns"]:
+                topic_name = next(iter(topic))
+                res_arn = sns.SNSNotification(ctx, imagename, region)._get_topic_arn(topic_name)
+
+                assert res_arn == f"arn:{partition}:sns:{region}:1234:{topic_name}"
```

Comment:
this will fail on the first missed res_arn. why not do something like:

```
expected = [f"arn:{partition}:sns:{region}:1234:{topic_name}" for topic in image_conf["sns"] for region in image_conf["regions"]
actual = []
for region ...
    for topic ...
       actual.append(...)
```
assert expected == actual

Reply:
Fixed the assert statement to compare the final result with the expected output.